online/src/main/scala/ai/chronon/online/FetcherModelUtils.scala

```diff
@@ -0,0 +1,171 @@
+    val futures = requestsByModel.map {
+      case (modelTransform, requests) =>
+        val identifiers = requests.map(_.modelRequestIdentifier)
+        val modelName = modelTransform.model.metaData.name
+        val mappedInputs = requests.map(_.inputs).map(modelTransform.mapInputs)
+        val joins = identifiers.map(_.joinName).distinct.sorted.mkString(",")
+        val ctx = Metrics.Context(environment = Metrics.Environment.ModelTransform, join = joins, model = modelName)
+        val startTs = System.currentTimeMillis()
+        ctx.increment(Metrics.Name.RequestCount)
+        val runModelInferenceRequest = RunModelInferenceRequest(modelTransform.model, mappedInputs)
+        val runModelInferenceResponseFuture = modelBackend.runModelInference(runModelInferenceRequest)
+        val modelResponses = runModelInferenceResponseFuture
+          .map { resp =>
+            val endTs = System.currentTimeMillis()
+            ctx.distribution(Metrics.Name.LatencyMillis, endTs - startTs)
+            resp
+          }
+          .map(_.outputs)
+          .map { outputs =>
+            outputs.zip(requests).map {
+              case (outputs, modelRequest) =>
+                val modelRequestIdentifier = modelRequest.modelRequestIdentifier
+                val mappedOutputs = modelTransform.mapOutputs(outputs)
+                ModelTransformResponse(modelRequestIdentifier, mappedOutputs, Map.empty)
+            }
+          }
+          .recover {
+            case e: Throwable =>
+              ctx.incrementException(e)
+              identifiers.map { identifier =>
+                ModelTransformResponse(identifier, Map.empty, Map(identifier.stringRep + "_exception" -> e))
+              }
+          }
+        modelResponses
+    }
+    Future.sequence(futures).map(_.flatten.toSeq)
+  }
+
+  /*
+   * Entry point to the ModelTransform stage, which takes in previous stage's output (derivations), run through
+   * model transforms via ModelBackend, and return the results
+   */
+  def fetchModelTransforms(joinRequestsFuture: Future[scala.collection.Seq[ResponseWithContext]],
+                           modelBackend: ModelBackend)(implicit
+      executionContext: ExecutionContext): Future[scala.collection.Seq[ResponseWithContext]] = {
+
+    val modelTransformsStartTs = System.currentTimeMillis()
+    joinRequestsFuture.flatMap { joinRequests =>
+      val hasModelTransforms = joinRequests.flatMap(_.joinCodec).exists(_.conf.hasModelTransforms)
+      if (!hasModelTransforms) {
+        Future.successful(joinRequests)
+      } else {
+        // Step1: For each join request, build a list of ModelTransformRequest, with ModelRequestIdentifier as join key
+        val modelTransformRequestsByJoinRequest = joinRequests.map { req =>
+          val modelTransformRequests = (req, buildModelTransformRequests(req))
+          val modelNames =
+            modelTransformRequests._2.map(_.modelRequestIdentifier.modelName).distinct.sorted.mkString(",")
```

Comment:
if we perform de-dup in `buildModelTransformRequests`, then we can remove `.distinct` here
```suggestion
            modelTransformRequests._2.map(_.modelRequestIdentifier.modelName).sorted.mkString(",")
```